src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyleDiagnosticAnalyzerBase.cs

```diff
@@ -46,50 +51,87 @@ private void CompilationStartAction(CompilationStartAnalysisContext context)
             var idToCachedResult = new ConcurrentDictionary<Guid, ConcurrentDictionary<string, string>>(
                 concurrencyLevel: 2, capacity: 0);
 
-            context.RegisterSymbolAction(c => SymbolAction(c, idToCachedResult), _symbolKinds);
+            context.RegisterSymbolAction(SymbolAction, _symbolKinds);
+            context.RegisterSyntaxNodeAction(SyntaxNodeAction, SupportedSyntaxKinds);
+            return;
+
+            // Local functions
+
+            void SymbolAction(SymbolAnalysisContext symbolContext)
+            {
+                var diagnostic = TryGetDiagnostic(
+                    symbolContext.Compilation,
+                    symbolContext.Symbol,
+                    symbolContext.Options,
+                    idToCachedResult,
+                    symbolContext.CancellationToken);
+
+                if (diagnostic != null)
+                {
+                    symbolContext.ReportDiagnostic(diagnostic);
+                }
+            }
+
+            void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
+            {
+                var diagnostic = TryGetDiagnostic(
+                    syntaxContext.Compilation,
+                    syntaxContext.SemanticModel.GetDeclaredSymbol(syntaxContext.Node, syntaxContext.CancellationToken),
```

Comment:
this feels strange.  one of the kinds you register is SyntaxKind.ForEachStatement... but GetDeclaredSymbol does'nt work on a ForEachStatement does it??

Reply:
yes it does

Reply:
it gives the symbol of the loop variable.. note that there isn't any other node you could get the loop variable from - it's not a VariableDeclaration/Declarator, it only has a single token for the variable name.